src/TestCase.js

```diff
@@ -14,12 +14,66 @@ const PLATFORM_EXCLUDED_RESOLUTIONS = [
 export const INVALID_REASON_REQUIRED = 'The required dimensions are missing';
 export const INVALID_REASON_RESOLUTION = 'This resolution is not available on this platform';
 
+class AbstractState {
+	finished;
+	description;
+
+	constructor() {
+		this.finished = false;
+	}
+}
+
+export class TestCasePendingState extends AbstractState {
+
+	constructor() {
+		super();
+		this.description = "Test case is pending"
+	}
+}
+
+export class TestCaseFinishedState extends AbstractState {
+	/**
+	 * @param {string} description
+	 */
+	constructor( description ) {
+		super()
+		this.finished = true;
+		this.description = description
+	}
+}
+
+export class TestCaseIsRunningState extends AbstractState {
+	/**
+	 * @param {string} description
+	 */
+	constructor( description ) {
+		super();
+		this.finished = false;
+		this.description = description;
+	}
+}
+
+export class TestCaseFailedState extends AbstractState {
+	error;
+
+	/**
+	 * @param {string} description
+	 * @param {Error} error
+	 */
+	constructor( description, error ) {
+		super();
+		this.description = description;
+		this.error = error;
```

Comment:
Technically, you're abusing js' lack of type safety here. This should be in the abstract class.

Reply:
I'm not seeing how I'm abusing type safety here - the using code can either rely on the interface of the `AbstractState`, or check for an instance of `TestCaseFailedState`, which is the only subclass of `AbstractState` where `error` makes sense.

I've made `error` type hint nullable to make sure it's still correctly typed with invalid test cases.